### ISSUE TITLE
[ENH] turning `numba` into a soft dependency

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#3567]
I think we should turn `numba` into a soft dependency, since it is not core to the framework, but used only in specific estimators.

---

key arguments for:

* the only key change is giving users the choice `numba` yes/no
* for users with `numba` installed, the `sktime` experience remains precisely the same
  * users without `numba` will be able to use `sktime`, currently they cannot
  * users without `numba` will still be able to find `numba` based estimators using `all_estimators`, similar to other soft dependencies
* it should be easier to upgrade `sktime` to newer python or package versions, of core dependencies, as `numba` is slow to update compatibility
* makes python 3.11 compatibility possible, at current date
* architectural: reduces coupling, increases layer separation; minimize soft dependencies, see context below

key arguments against:

I don't really see any except the work required

---

Update: more context on architectural and usage benefits.

the general desideratum is "minimize dependencies". This is an explicit architectural north star of scikit-learn as well as sktime.

See e.g., this paper on sklearn design,
https://mblondel.org/publications/lbuitinck-ecmlpkdd2013.pdf
search for "dependencies" in the text.

In the past 6 months, we have made the interfaces of estimators much clearer, including that we can manage python versions and package dependencies on the estimator level, rather than at the sktime package level

This has prompted multiple lines of refactoring around "minimize dependencies":

* migrate sktime-dl into sktime because we can isolate tensorflow to estimators
* isolate `statsmodels`
* isolate `numba`
* newly added annotation module soft dependencies are fully isolated

`statsmodels` isolation and soft dep is very similar, see this PR: https://github.com/sktime/sktime/pull/3821

---

FAQ

> does this remove `numba` support from `sktime`?

No, `numba` will still be supported and used in estimators.
It will only change to be optional, for people who do not want (or need) to use estimators that depend on `numba`.

> will this impact performance of `sktime`?

No. For users who have `numba` installed, the experience is precisely the same. The logic of `numba` based estimators stays the same.

No change in estimator accuracy or speed. (evidence: look at the tests in https://github.com/sktime/sktime/pull/3843 which test for both)

The only change is that users get a choice whether they want `numba` or not.

> How can this be implemented within the estimators? Numba requires tags/decorators at each function to be compiled.

See https://github.com/sktime/sktime/pull/3631

strategy is the same as any dependency isolation on estimator level and uses the `sktime` base class layer:

* move `numba` imports into functions and classes, where possible
* add the package dependency tag to `numba` dependent estimators to isolate them
* use `pytest.skipif` to conditionally skip `numba` dependent non-suite tests if `numba` is not installed
* specific to `numba` due to use of decorators: where loose functions are decorated with `jit` and `njit`, move them to a private module `_orig_modulename_numba`, and move imports of that into functions or classes

Same as isolating any soft dependency, see, e.g., same strategy for `statsmodels`:

https://github.com/sktime/sktime/pull/3445
https://github.com/sktime/sktime/pull/3515
https://github.com/sktime/sktime/pull/3632
https://github.com/sktime/sktime/pull/3821

**Comment by patrickzib** [comment]
How can this be implemented within the estimators? Numba requires tags/decorators at each function to be compiled.

**Comment by patrickzib** [comment]
For example, how would you implement a soft dependency for this method:

```python
@njit(
    "float32[:](float32[:,:],int32[:],int32[:],float32[:],optional(int32))",
    fastmath=True,
    parallel=False,
    cache=True,
)
def _fit_biases(X, dilations, num_features_per_dilation, quantiles, seed):
   ...
```

**Comment by MatthewMiddlehurst** [comment]
I think this would have a large impact on the classification and clustering packages. There are probably 20+ files in sktime so far which use it.

**Comment by TonyBagnall** [comment]
also used in the distances, transformations and regression package. I do not see the benefit

**Comment by fkiraly** [comment]
it's mostly localized, and minimizing soft dependencies is of general benefit

**Comment by fkiraly** [comment]
>For example, how would you implement a soft dependency for this method:

Simplest way:

```python
if _check_soft_dependencies("numba", severity="none"):
    @njit(
        "float32[:](float32[:,:],int32[:],int32[:],float32[:],optional(int32))",
        fastmath=True,
        parallel=False,
        cache=True,
    )
    def _fit_biases(X, dilations, num_features_per_dilation, quantiles, seed):
       ...
```

**Comment by TonyBagnall** [comment]
Adds a large amount of boilerplate code, I'm against this

**Comment by patrickzib** [comment]
I second @TonyBagnall 's opinion.

**Comment by fkiraly** [comment]
The reason for the large amount of boilerplate code is that many of the distances and utility functions are lying around loosely and not part of any class. Were they part of an object, we could deal with this via a tag or import-on-contsruction.

**Comment by ngupta23** [comment]
> I think we should turn `numba` into a soft dependency, since it is not core to the framework, but used only in specific estimators.
> 
> Any thoughts?

This would be very helpful. This is one of the reasons why we are unable to upgrade to the latest version of `sktime` since `numba` has some problems in our environment. If it could be made optional, it would be awesome!

Ref: https://github.com/numba/numba/issues/4908
cc: @hancelpv

**Comment by MatthewMiddlehurst** [comment]
Do you have any more information about this installation issue @ngupta23? i.e. the Python and numba versions. Whether numba is made a soft dependency or not, it is still an issue to resolve.

**Comment by patrickzib** [comment]
Agreed. you would not be able to use around 50% (?) of all time series classification, regression, transformation packages without numba being installed.

**Comment by chrisholder** [comment]
As an additional point, I think we should be using numba as often as possible (where appropriate) for our estimators. It offers significant performance gains when used correctly at almost no detriment to user or developer. 

By making it a soft dependency we're implying it's a niche utility. It shouldn't be perceived this way but instead as one of sktimes developers best ways to optimise their code.

**Comment by fkiraly** [comment]
> Agreed. you would not be able to use around 50% (?) of all time series classification, regression, transformation packages without numba being installed.

Well, some users are not able to use 100% of all time series classification, regression, and transformation estimators because numba is installed, see above @ngupta23's comment.

We should strive to minimize dependencies and make optional parts of `sktime` optional, not force users install packages they do not really need.

**Comment by fkiraly** [comment]
> By making it a soft dependency we're implying it's a niche utility. It shouldn't be perceived this way but instead as one of sktimes developers best ways to optimise their code.

Don't think that is the implication at all.

I think it just implies that we are managing our interface points well, and minimize coupling where we can.
https://stackoverflow.com/questions/39946/coupling-and-cohesion

**Comment by fkiraly** [comment]
(we can probably get to a point where `pandas` and `scikit-learn` are the only direct dependencies!)

**Comment by patrickzib** [comment]
> > Agreed. you would not be able to use around 50% (?) of all time series classification, regression, transformation packages without numba being installed.
> 
> Well, some users are not able to use 100% of all time series classification, regression, and transformation estimators because numba is installed, see above @ngupta23's comment.
> 
> We should strive to minimize dependencies and make optional parts of `sktime` optional, not force users install packages they do not really need.

Numba should be a must for the functionailty provided in sktime. It is the obvious way to get best performance. 

> I don't think it is too relevant whether the combined forces of the four of you can fix @ngupta23's issue, since it is just a representative for interactions/restrictions that are caused by having more core dependencies.

You should not fix a bug by removing 50% of a libraries functionality but by addressing the issue

**Comment by ngupta23** [comment]
> Whether numba is made a soft dependency or not, it is still an issue to resolve.

Can you add this to the sktime setup? Thanks we can retry this to see if it resolves the issue?

https://github.com/numba/numba/issues/4908#issuecomment-571537473

```
Another workaround is to add zip_safe=False to the setup.py file of the packages that use numba caching.
```

**Comment by MatthewMiddlehurst** [comment]
If you can provide the system/package versions of the failed install attempt i.e.
```
System:
    python: 3.8.10 | packaged by conda-forge | (default, May 11 2021, 07:01:05)  [GCC 9.3.0]
executable: /opt/conda/bin/python
   machine: Linux-5.4.0-109-generic-x86_64-with-glibc2.10

Python dependencies:
          pip: 22.1
   setuptools: 49.6.0.post20210108
      sklearn: 0.24.2
       sktime: 0.13.4
  statsmodels: 0.13.2
        numpy: 1.22.4
        scipy: 1.6.3
       pandas: 1.4.2
   matplotlib: 3.5.2
       joblib: 1.0.1
        numba: 0.56.3
     pmdarima: None
      tsfresh: 0.19.0
```
or similar, I can try to reproduce it locally then fix it in a PR. The thread you linked has a few ideas to try, thanks.

I have never encountered an install failure due to numba, so I imagine there are some specific circumstances.

**Comment by ngupta23** [comment]
It was with sktime 0.12.0. I can not produce the issue locally either. It is specific to our production environment where write access may be limited. 

```
Error:
RuntimeError: cannot cache function 'mean': no locator available for file '/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/utils/numba/stats.py'
```

This is similar to the issue that I linked in the numba repo.

```
python=3.7.10 on Linux

- sktime=0.12.0
- pip=22.0.3
- setuptools=60.9.0
- scikit-learn=1.0.2
- statsmodels=0.13.2
- numpy=1.21.5
- scipy=1.7.3
- pandas=1.1.5
- matplotlib-base=3.5.1
- joblib=1.1.0
- numba=0.55.2
- pmdarima=1.8.5
- tsfresh=0.19.0
```

**Comment by patrickzib** [comment]
Could it be related to this bug report
https://github.com/numba/numba/issues/4908 ?

E.g. installing as egg. A solution is listed as:
 Add zip_safe=False to the setup.py of the packages using Numba caching:

**Comment by ngupta23** [comment]
> E.g. installing as egg. A solution is listed as:
> Add zip_safe=False to the setup.py of the packages using Numba caching:

That is what I recommended to be added to sktime to see if the issue is resolved. (https://github.com/sktime/sktime/issues/3567#issuecomment-1285987805)

**Comment by ngupta23** [comment]
Actually, I just checked, and it seems to be there already. So, I am not sure if that would help.

https://github.com/sktime/sktime/blob/v0.12.0/setup.py#L65

**Comment by MatthewMiddlehurst** [comment]
From the original https://github.com/numba/numba/issues/4908 issue: "Another workaround is to install the package with pip install ., which will not create an egg, instead of python setup.py install."

Did both of these fail? Can try reproducing at some point with your settings, but sounds like it will be a pain to fix 🙂

**Comment by ngupta23** [comment]
We are already using pip for `sktime`. Here is the error stack if it helps. The error comes when we try to import any of the forecasting classes/functions such as `make_reduction`, `ThetaForecaster`, etc. Under the hood, it seems to be calling classification code that used numba...


```python-traceback
cannot cache function 'mean': no locator available for file '/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/utils/numba/stats.py'Traceback (most recent call last):
  File "/hadoop/yarn/usercache/livy/appcache/application_1666359845927_0004/container_e07_1666359845927_0004_01_000001/__pyfiles__/user_def.py", line 16, in user_def
    from sktime.forecasting.compose import make_reduction
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/forecasting/compose/__init__.py", line 36, in <module>
    from sktime.forecasting.compose._pipeline import (
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/forecasting/compose/_pipeline.py", line 12, in <module>
    from sktime.registry import scitype
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/registry/__init__.py", line 4, in <module>
    from sktime.registry._base_classes import (
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/registry/_base_classes.py", line 64, in <module>
    from sktime.classification.early_classification import BaseEarlyClassifier
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/classification/early_classification/__init__.py", line 9, in <module>
    from sktime.classification.early_classification._probability_threshold import (
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/classification/early_classification/_probability_threshold.py", line 20, in <module>
    from sktime.classification.interval_based import CanonicalIntervalForest
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/classification/interval_based/__init__.py", line 11, in <module>
    from sktime.classification.interval_based._cif import CanonicalIntervalForest
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/classification/interval_based/_cif.py", line 18, in <module>
    from sktime._contrib.vector_classifiers._continuous_interval_tree import (
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/_contrib/vector_classifiers/_continuous_interval_tree.py", line 21, in <module>
    from sktime.utils.numba.stats import iqr, mean, median, numba_max, numba_min, slope, std
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/utils/numba/stats.py", line 8, in <module>
    @njit(fastmath=True, cache=True)
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/numba/core/decorators.py", line 212, in wrapper
    disp.enable_caching()
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/numba/core/dispatcher.py", line 863, in enable_caching
    self._cache = FunctionCache(self.py_func)
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/numba/core/caching.py", line 601, in __init__
    self._impl = self._impl_class(py_func)
  File "/opt/conda/envs/<hidden>/lib/python3.7/site-packages/numba/core/caching.py", line 338, in __init__
    "for file %r" % (qualname, source_path))
RuntimeError: cannot cache function 'mean': no locator available for file '/opt/conda/envs/<hidden>/lib/python3.7/site-packages/sktime/utils/numba/stats.py'
```

**Comment by MatthewMiddlehurst** [comment]
This doesn't seem to be a setup or installation problem similar to the issue linked IMO. Have you tried changing the numba cache directory to a location where you do have write access? An example used is `ENV NUMBA_CACHE_DIR=/tmp/numba_cache` but it could be any writable location really. Perhaps we should set something up where write access is limited similar to this @TonyBagnall.

It would be hard to test, but `BaseEarlyClassifier` is in the same init.py as the actual classifiers which import numba functions, which is where the numba is coming from in the error. Putting it in its own base directory or with `BaseClassifier` will stop numba being imported from base class imports, perhaps.

**Comment by fkiraly** [comment]
@patrickzib, I think your responses do not follow.

>Numba should be a must for the functionailty provided in sktime. It is the obvious way to get best performance. 

This does not follow, the two statements are not logically linked. If you want to use the estimators, you install `numba`, just like with any other soft dependencies.

> You should not fix a bug by removing 50% of a libraries functionality but by addressing the issue

It's not 50% of the library.
And which issue are you referring to?

**Comment by patrickzib** [comment]
Ok, @fkiraly . Let us have a vote on turning numba into a soft dependency then. Everyone gave his opinion. Obviously, there are different views on that and no one is moving.

**Comment by patrickzib** [comment]
I am against this. Let us have a vote.

**Comment by patrickzib** [comment]
As I see it: numba is not a nice to have for most modules in classification/regression/clustering/transformers that currently use it. No matter if it is 50% or 30% or 10%. All SotA modules like Rocket, etc. use it.

If you turn numba off, there will be a slowdown of a factor of 10-100. Especially, if the users are not aware that they must install numba manually to get best performance.

**Comment by MatthewMiddlehurst** [comment]
The estimators which use numba will simply be unavailable to use while numba is not installed. I'm slightly sceptical of the feasibility of isolating it all, however.

**Comment by patrickzib** [comment]
@MatthewMiddlehurst Agreed. 

Let me rephrase it: due to the large speedup numba offers, we should be working on refactoring more modules into using numba - rather than adding large amounts of boilerplate code to existing modules using numba.

**Comment by fkiraly** [comment]
it seems like `numba` is the only offending core dependency for 3.11 support.

**Comment by fkiraly** [comment]
voting has started on the core developer channel, @patrickzib called it on #3631 and @aiwalter has set it up, so we will proceed by the rules

I'm on the "for" side, based on:

* "minimize dependencies" is a core design principle of sktime
* numba install is creating problems for a lot of our new users in my view of people asking for help, so it might cause frustration/churn in the trial phase if people don't get `sktime` installed
* numba is currently the only core dep that hasn't figured out python 3.11 (afaik), so imo likely will keep us back there
* it is localized to estimators/baseobjects, so could be turned into a soft dep along the lines of other estimators with a soft dep
* makes architectural sense since numba is not required in the framework, "minimize coupling, maximize cohesion"
* our most popular reverse dep, pycaret, is complaining about numba not playing well in their prod system, so they are stuck on sktime 0.12 (see above @ngupta23's posts)
* I don't think the X% estimators thing is a valid argument. The experience for people who manage to get `numba` installed with `sktime` would be exactly the same. The difference in experience would be for the people who don't want to (or cannot) get `numba` installed.

**Comment by ngupta23** [comment]
> * our most popular reverse dep, `pycaret`, is complaining about numba not playing well in their prod system, so they are stuck on sktime 0.12 (see above @ngupta23's posts)

Just want to clarify that this is unrelated to `pycaret`. This is for work and we are actually stuck at sktime 0.7.0. Anything above that starts to fail due to `numba`.

**Comment by TonyBagnall** [comment]
I dont think you should group people together as @TonyBagnall et al

**Comment by TonyBagnall** [comment]
I am against it because I think it will result in code bloat, I'm not convinced it will actually work as a soft dependency and I think a larger section of the code base than you realise will not work without it. Bur I am not going to champion the against view, I will work with what you come up with

**Comment by patrickzib** [comment]
To use numba simply add the tag `@nit` to a python method:

```python
def distance(x, y):
    d = 0.0
    for (a,b) in zip(x,y):
        d = u + (a - b)**2
    return d

@njit
def distance_numba(x, y):
    d = 0.0
    for (a,b) in zip(x,y):
        d = u + (a - b)**2
    return d

x = np.random.random((10_000_000))
y = np.random.random((10_000_000))
d = distance(x, y)
d = distance_numba(x, y)
```

The runtime difference between these two methods is striking:

> Time distance:  3.925307035446167 (seconds)
> Time distance_numba:  0.08873510360717773 (seconds)

**There is factor of 45 difference in runtimes!** 

IMHO: We should spent time more time on refactoring existing modules into using numba

**Comment by fkiraly** [comment]
>soft or hard dependency, we have to support it anyways if a bug is reported

correct, but that's neither a pro or a con, since it is the same in both cases

> “minimizing dependencies” will firstly disable all modules using numba for most “beginner” users of sktime, and we will have to write long guidelines how to install numba for these

Here's the long guideline:

`pip install numba`

> turning numba into a soft-dependency will add large amounts of boilerplate code to every module, and every method therein, using it
    around 50% (a best guess) of all sktime classifiers/regressors/transformers use numba

I don't think so?

Best case, we are looking at adding a tag and a little bit of refactor

> You do not fix a bug by hiding large parts of a library but by solving the underlying issue

I think this is misconception/misrepresentation.
There is no bug. We are talking about a refactor.

**Comment by fkiraly** [comment]
just for the record, I'm *in favour* of making `numba` a soft dep mid-term, for the vote

**Comment by fkiraly** [comment]
> IMHO: We should spend more time on refactoring existing modules into using numba rather than hiding it behind a soft dependency

False alternatives. It is not either or.

It seems to me you do not fully understand what soft dependency means, or I don't understand what you mean with "hiding", and why it would prevent us from refactoring towards `numba`.

**Comment by patrickzib** [comment]
@fkiraly  you are contradicting yourself. You are claiming numba is hard to install and maintain:

> numba install is creating problems for a lot of our new users in my view of people asking for help, so it might cause frustration/churn in the trial phase if people don't get sktime installed

Yet you say:

> Here's the long guideline:
> pip install numba

**So clearly, if it is that easy, why are we having this PR?**

> Best case, we are looking at adding a tag and a little bit of refactor

Ok, please rewrite Rocket to prove my statement wrong. Then we can see, if the added code is really that little.

**Comment by patrickzib** [comment]
There are two kinds of users, I claim:

- Forecatsting: the might not need numba
- Transformers/Segmentation/Distances/Classfiication/Regression/Clustering: they require numba to be installed along sktime to use it

**Comment by patrickzib** [comment]
> fkiraly: I think this is misconception/misrepresentation.
> There is no bug. We are talking about a refactor.

You claimed there are issues (bug reports) in using sktime along numba:

> fkiraly: our most popular reverse dep, pycaret, is complaining about numba not playing well in their prod system, so they are stuck on sktime 0.12 (see above @ngupta23's posts)

And your presented solution is to remove the dependency to numba. 

Thus my statement from before:

> patrickzib: You do not fix a bug by hiding large parts of a library but by solving the underlying issue

**Comment by TonyBagnall** [comment]
fkiraly: our most popular reverse dep, pycaret, is complaining about numba not playing well in their prod system, so they are stuck on sktime 0.12 (see above @ngupta23's posts)

just to clarify this point for any one picking up this thread from here, @ngupta23 says this is not pycaret problem, but a problem with a commercial product that is unrelated to sktime afaik. 

Also, given this is a vote on medium term design principle not a concrete PR, should the vote not also include statsmodels and anything else not numpy or pandas? Do you want scipy as a soft dependency too @fkiraly ?

**Comment by ngupta23** [comment]
> but a problem with a commercial product that is unrelated to sktime afaik.

But the issue is that we cannot install sktime due to numba hard dependency which is not needed for forecasting at all as far as I can see.

**Comment by patrickzib** [comment]
> But the issue is that we cannot install sktime due to numba hard dependency which is not needed for forecasting at all as far as I can see.

Very much appriciated. And @MatthewMiddlehurst is working on it:
https://github.com/sktime/sktime/pull/3634#issuecomment-1290858158

**Comment by fkiraly** [comment]
#3634 does not resolve what @ngupta23 said, that a hard dependency is not needed for forecasting.

**Comment by patrickzib** [comment]
> #3634 does not resolve what @ngupta23 said, that a hard dependency is not needed for forecasting.

Imo, it would be better to avoid thinking in terms of "forecasting" and "classification". sktime is a library for all users, and none should be prioritized.

`statsmodel` is not needed for classification. Yet, it is imported, too, and causes a lot of problems regularly.

**Comment by fkiraly** [comment]
> `statsmodel` is not needed for classification. Yet, it is imported, too, and causes a lot of problems regularly.

Yes, and that's why we should seriously consider making `statsmodels` a soft dependency.

> Imo, it would be better to avoid thinking in terms of "forecasting" and "classification/others". sktime is a library for all users, and none should be prioritized.

Fully agree!
That's why I think we should not have "module specific soft dependencies".

**Comment by fkiraly** [comment]
I have found a better way to isolate `numba` than the if check:

* move all `numba` decorated functions into private modules
* move imports for all direct numba calls and use of those functions into `_fit`, `_predict`, etc
* import functionality from private modules in estimators

That should work and be less clunky.

**Comment by MatthewMiddlehurst** [comment]
Since the discussion around the issue seen by @ngupta23 is here, I will post this here also.

#3799 should fix the issue of the following code snippet failing, at least. Version 0.14.1 is being released soon, so it should go in with that.

```
from sklearn.preprocessing import MinMaxScaler
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import ForecastingPipeline
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.naive import NaiveForecaster
from sktime.transformations.series.adapt import TabularToSeriesAdaptor
from sktime.transformations.series.impute import Imputer

if __name__ == "__main__":
    y, X = load_longley()
    y_train, _, X_train, X_test = temporal_train_test_split(y, X)
    fh = ForecastingHorizon(X_test.index, is_relative=False)
    pipe = ForecastingPipeline([
        Imputer(method="mean"),
        TabularToSeriesAdaptor(MinMaxScaler()),
        NaiveForecaster(strategy="drift"),
    ])
    pipe.fit(y_train, X_train)
    y_pred = pipe.predict(fh=fh, X=X_test)
    print(y_pred)
```

If that is not sufficient and people are still seeing issues, this could possibly work from the user side. It is a bit of a nuclear option, but placing it at the top of your file may also work.

```
from numba import jit, config
config.DISABLE_JIT = True
```

numba has an environmental variable to disable compilation of numba functions https://numba.pydata.org/numba-doc/dev/reference/envvars.html#envvar-NUMBA_DISABLE_JIT. 

I wouldn't recommend using this in normal circumstances. Currently, disabling JIT will have no impact on the forecasting module as far as I am aware. However, if the dependency of a wrapped forecaster uses numba the efficiency may be impacted, I don't know enough about the module to tell you if any do. This will also make other code which uses numba orders of magnitude slower, potentially.

**Comment by patrickzib** [comment]
Closed as per majority vote of core developers.

**Comment by fkiraly** [comment]
please don't close issues that are unresolved

**Comment by patrickzib** [comment]
Following the same idempotent logic as outlined here, I hereby start proposing additional issues to remove soft dependencies not needed by sktime to run estimators and reduce hard dependencies: 
https://github.com/sktime/sktime/issues/3873

**Comment by patrickzib** [comment]
- Turning Pandas into a soft dependency: https://github.com/sktime/sktime/issues/3874

**Comment by patrickzib** [comment]
- Turning scipy into a soft dependency: https://github.com/sktime/sktime/issues/3875

**Comment by patrickzib** [comment]
We should create an overall issue to track all hard dependencies that must be turned into a soft dependency.

**Comment by patrickzib** [comment]
I think this thread became really hard to follow with all the changes over the times. Thus, please use this new thread for all arguments for or against in the future: 
- turning all non-core libraries into a soft dependency https://github.com/sktime/sktime/issues/3876

It aggregates the current state of discussions.

**Comment by ngupta23** [comment]
> Since the discussion around the issue seen by @ngupta23 is here, I will post this here also.
> 
> #3799 should fix the issue of the following code snippet failing, at least. Hopefully no other forecasters have this issue, please post if they do. Version 0.14.1 is being released soon, so it should go in with that.

I wanted to provide an update to this. First of all, thank you to everyone who worked on isolating the calls to numba. I can confirm that with the 0.14.1 release, we were able to use the Forecasting code on our platform without running into numba compilation issues. Really appreciate the work that went into making it happen. Now to summarize why it worked for us right now.

**Our previous environment**
* Contained `numba`
* Contained `sktime` < 0.14.1

Since our environment contained `numba` already, just making `numba` a soft dependency in `sktime` would not have resolved the issue. The root cause in our case was that there are some write permission issues because of which `numba` could not compile functions. But when using the forecating code only (as in our case), `numba` should not have been called in the first place since none of the forecasting code relies on `numba`. Hence, the current fix of isolating `numba` in the sktime library (to be called when absolutely needed) worked since we are using only Forecasting code. 

In the future though, if we ever want to use classification from sktime, we will run into the same numba compiling issues unless we fix the write permission issues on our end. But that is not a sktime problem :).

Thanks again to everyone who worked on this fix.

**Comment by fkiraly** [comment]
@ngupta23, you're welcome!

Regarding the "in the future" scenario, it would then probably be beneficial to isolate `numba` also inside the classification module? I.e., not require any `numba` imports except if you want to use a `numba` dependent classifier?

The topic is currently a bit controversial internally, but I'll just keep this experimental PR updated with `main`, so it can be used as a `numba` isolation patch: https://github.com/sktime/sktime/pull/3843

**Comment by fkiraly** [comment]
`numba` is a soft dependency since 0.18.0 now.
